doc/dev/crimson/poseidonstore.rst

```diff
@@ -0,0 +1,401 @@
+      - NVMe provides atomicity guarantees for a write command (Atomic Write Unit Power Fail)
+        For example, 512 Kbytes of data can be atomically written at once without fsync()
+      - Small size (object_info_t, snapset, collection, etc.) can be embed
+      - stage 1
+        WAL (written) --> | TxBegin A | Log Entry | TxEnd A | (if small)
+        Data partition (written)--> | Data blocks | (if large)
+    - 2. Then, append the data to WAL (if large)
+      - stage 2 (Updating object meta and allocation bitmap can be skipped via pre-allocation)
+        WAL --> | TxBegin A | Log Entry | TxEnd A | (if large)
+
+- Read
+  - Use the cached object metadata to find out the data location
+  - If not cached, need to search WAL after checkpoint and Object meta partition to find the 
+    latest meta data
+
+- Flush
+  - Flush WAL entries whenever needed
+
+
+Crash consistency
+------------------
+Large case
+- 1. Crash occurs right after writing Data blocks
+  - Data partition --> | Data blocks |
+  - We don't need to care this case. Data is not alloacted yet in reality. The blocks will be reused.
+- 2. Crash occurs right after WAL using atomic write command
+  - WAL --> | TxBegin A | Log Entry| TxEnd A |
+  - Data partition --> | Data blocks |
+  - Write procedure is completed, so there is no data loss or inconsistent state
+
+Small case
+- 1. Crash occurs right after writing WAL
+  - WAL --> | TxBegin A | Log Entry| TxEnd A |
+  - All data has been written
+
+
+Comparison
+----------
+- Best case (pre-allocation)
+  - Only need two writes on both WAL and Data partition without updating Object meta and Data bitmap
+- Worst case (flush happens)
+  - Without embedding, at least three writes are required additionally on Object meta, Data bitmap, and Data blocks
+- WAL needs to be flushed if the WAL is close to full and the WAL entry is snapset, object_info_t, 
+  collection and pg_log
+  - The premise behind this design is OSD can manage the lateset metadta as a single copy. So,
+    appended entires are not to be read
+- Either best of worst case does not produce severe I/O amplification (it produce I/Os, but I/O rate is constant) 
+  unlike LSM-tree DB (proposed design is similar to LSM-tree which has only level-0)
+  
+
+Detailed Design 
+===============
+
+- Onode lookup
+ - Flat namespace
+  Object data, object meta, omap, and xattrs are all the data related to an object (we call it a head object from ghobject).
+  PoseidonStore adds different suffixes to the head object ID to create PSD_OID to distinguish them
+  (e.g., Assuming head object ID = 12.e4 -> Object data's PSD_OID = 12.e4_data, Object meta's PSD_OID = 12.e4_meta, omap's 
+  PSD_OID = 12.e4_omap_key1, xattr's PSD_OID = 12.e4_xattr_key1).
+  By eliminating suffixes, we can recover their head object ID. Thus, we can know they are all related to the same object.
+  Note that because all objects are sorted by their PSD_OIDs in lexicographic order, xattrs and omap can be iterated 
```

Comment:
This has a disadvantage: scanning objects for backfill and scrub will require enumerating omap/xattr data.  How do you plan on representing an object with no xattrs or omap?